File: 0740-delete-and-earn/0740-delete-and-earn.cpp
class Solution {
private:
    int solve(int i, unordered_map<int,int>& hashMap, vector<int>& memo) {
        // Base cases
        if (i == 0) return 0;
        if (i == 1) return hashMap[1];
        
        // Check memo
        if (memo[i] != -1) return memo[i];
        
        // Same recurrence: max(dp[i-1], dp[i-2] + hashMap[i])
        int notChoose = solve(i-1,hashMap,memo); 
        int choose = solve(i-2,hashMap,memo) + hashMap[i]; 
        memo[i] = max(choose, notChoose);
        return memo[i];
    }
    
public:
    int deleteAndEarn(vector<int>& nums) {
        sort(nums.begin(), nums.end()); 
        unordered_map<int,int> hashMap; 
        int max_ = *max_element(nums.begin(), nums.end());
        for(int n : nums) hashMap[n] += n; 
        vector<int> memo(max_ + 1, -1); 
        
        return solve(max_, hashMap, memo);
    }
};

File: 0740-delete-and-earn/0740-delete-and-earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        if(nums.length == 1) return nums[0]; 
        Map<Integer,Integer> hashMap = new HashMap<>(); 
        int max_ = 0; 
        for(int n : nums){
            max_ = Math.max(max_, n); 
            hashMap.put(n, hashMap.getOrDefault(n,0) + 1); 
        }

        int[] dp = new int[max_ + 1]; 
        dp[1] = 1 * hashMap.getOrDefault(1,0); 

        int answer = dp[1]; 
        for(int i = 2; i <= max_; i++){
            dp[i] = Math.max(dp[i-2] + i * hashMap.getOrDefault(i,0), dp[i-1]); 
            answer = Math.max(answer, dp[i]);
        }
        return answer; 

    }
}